R/site.R

```diff
@@ -0,0 +1,191 @@
+#' Render multiple documents as a website
+#'
+#' Render all of the R Markdown documents within a directory as a website.
+#'
+#' @param input Website directory (or the name of a file within the directory)
+#' @param output_format R Markdown format to convert to.
+#' @param quiet \code{TRUE} to supress messages and other output.
+#' @param encoding The encoding of the input file; see \code{\link{file}}.
+#' @param ... Currently unused
+#'
+#' @details
+#'
+#' To render a group of R Markdown documents within a directory as a website
+#' there are two requirements:
+#' \itemize{
+#'   \item{Include a file named \code{index.Rmd} which will serve as the site's
+#'   home page.}
+#'   \item{Add a \code{site: true} entry to the YAML metadata of \code{index.Rmd}}
+#' }
+#'
+#' Once you've met these requirements you can call the \code{render_site} function
+#' on the directory (or any Rmd file within the directory) to render all of the
+#' R Markdown documents within the directory.
+#'
+#' @section Custom Site Generation:
+#'
+#' The default site generation function (\code{rmarkdown::default_site}) simply
+#' renders all of the files in the input directory. You can however define a
+#' custom site generator which has alternate behavior. There are two ways for
+#' a custom site generator to be bound to from within \code{index.Rmd}:
+#'
+#' \itemize{
+#'   \item{The active output format defines a "_site" function (e.g.
+#'   \code{html_document_site}).}
+#'   \item{The \code{site} metadata entry refers to a custom function (e.g.
+#'   \code{mypackage::site}).}
+#' }
+#'
+#' The site generation function should return a list with the following
+#' elements:
+#'
+#' \itemize{
+#'   \item{\code{name}} {The name for the website (e.g. the parent directory name).}
+#'   \item{\code{output_dir} {The directory where the website output is written to
+#'   (e.g. "." or "_site")}}
+#'   \item{\code{render}} {An R function that can be called to generate the site.
+#'   The function should accept the \code{output_format}, \code{quiet},
+#'   \code{encoding}, and \code{...} arguments}
+#' }
+#'
+#' See the source code of the \code{rmarkdown::default_site} function for a
+#' simple example of a custom site generation function.
+#'
+#' @section Site Preview:
+#'
+#' Note that the \code{render_site} function can be used as a custom \code{knit}
+#' handler within RStudio by adding the following to the YAML of any
+#' document within a site directory:
+#'
+#' \code{knit: rmarkdown::render_site}
+#'
+#' This will result in the entire site being rendered each time Knit is
+#' executed within RStudio.
+#'
+#' @export
+render_site <- function(input = ".",
+                        output_format = NULL,
+                        quiet = FALSE,
+                        encoding = getOption("encoding"),
+                        ...) {
```

Comment:
I think we need an argument `envir = parent.frame()` here, and pass it all the way to the site generator, otherwise `rmarkdown::render()` will be using an environment inside a certain function.